### PR TITLE
docs: Document Yarn support in cachi2

### PR DESCRIPTION
**Pull request by eskultety**
# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [n/a] Code coverage from testing does not decrease and new code is covered
- [x] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -362,6 +362,20 @@ Make sure lockfile version is higher than v1 (Node.js 15 or higher).
...
+### yarn
+
+<https://v3.yarnpkg.com/>
+
+Current version: v3
+
+Cachi2 supports Yarn by parsing [package.json](https://v3.yarnpkg.com/configuration/manifest/)
+file present in the source repository and downloading the declared dependencies.
+
+If not checked in to the repository, a lockfile (``yarn.lock``) will automatically be generated by
```
Comment (brunoapimentel) [inline README.md:374]: I did a few tests, and the prefetch fails when the yarn.lock file is not present:

```
Internal Error: sample-nodejs-app@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
```

Maybe we should wrap this into a nicer error? In any case, I think to keep the current ethos of Cachi2, we shouldn't allow a repo without a lockfile to be processed.

Comment (ben-alkov) [inline README.md:374]:
```suggestion
If not already checked in to the repository, a lockfile (``yarn.lock``) will automatically be generated by
```

**Comment by brunoapimentel** [comment]
I think we also want to have a complete example, following what has been done on other package managers: https://github.com/containerbuildsystem/cachi2/blob/main/docs/usage.md#usage-examples.

I'm ok if this is done in a follow-up PR, though.

File: README.md (hunk 1 above)
Comment (eskultety) [reply README.md:374]: I'll drop this sentence with https://github.com/containerbuildsystem/cachi2/pull/425#discussion_r1426765531 in mind.

**Comment by eskultety** [comment]
> I think we also want to have a complete example, following what has been done on other package managers: https://github.com/containerbuildsystem/cachi2/blob/main/docs/usage.md#usage-examples.
> 
> I'm ok if this is done in a follow-up PR, though.

Nope, that's a good point which needs to be addressed in this PR.

File: docs/yarn.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+### Dealing with .yarnrc.yml
+Cachi2 parses the project's ``.yarnrc.yml`` file and analyzes configuration settings. Before cachi2
+proceeds with the actual dependency fetching, it verifies whether all [configuration
+settings](https://v3.yarnpkg.com/configuration/yarnrc) that set a path to a resource don't point
+outside of the source repository, so in order to avoid any issues reported by Cachi2 in this regard
+make sure all your project resource references are bound by the repository. Part of the analysis of
+the repository's ``.yarnrc.yml`` file is detection of plugin usage which is further explained in
+[Dealing with plugins](#dealing-with-plugins).
+
+### Dealing with Yarn Zero-Installs
```
[the remark this answers is not in the document]
Comment (eskultety) [reply docs/yarn.md:47]: > I'm not sure what to do in this case, though, I just think it ends up being bad UX, besides the legitimate concern of forbidding zero-installs. 

I have an idea on how to fix this that I'll try to materialize next year, but for the purposes of the Yarn support release, covering this in the docs for now shouldn't be a blocker for the release, right? Or do you think it's something critical?

File: README.md (hunk 1 above)
Comment (eskultety) [reply README.md:374]: Reworded the docs accordingly, but a fix will have to be posted for the issue.

File: docs/yarn.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (eskultety) [reply docs/yarn.md:48]: I added a _Known pitfalls_ section that explains the matter, let me know if it's enough for the time being or whether we need to fix this before the release.

**Comment by eskultety** [comment]
> I think we also want to have a complete example, following what has been done on other package managers: https://github.com/containerbuildsystem/cachi2/blob/main/docs/usage.md#usage-examples.

Added the change, however, @brunoapimentel I'd like to ask you to move your sample-nodejs-app Yarn modifications to the main cachito nodejs repo which my docs changes refer to. Also, I think you'll need to do a few cosmetic changes there so that the docs matches is, e.g. the Dockerfile.

**Comment by eskultety** [comment]
Marking this as a draft since the docs contains information and steps that assume Yarn to be enabled which it's not yet. This PR can still be reviewed in the meantime though, we should only postpone its merging until the very last moments before releasing support for Yarn v3.

File: README.md [hunk 3]
```diff
@@ -362,6 +362,20 @@ Make sure lockfile version is higher than v1 (Node.js 15 or higher).
 
 See [docs/npm.md](docs/npm.md) for more details.
 
+### yarn
+
+<https://v3.yarnpkg.com/>
+
+Current version: v3
+
+Cachi2 supports Yarn by parsing [package.json](https://v3.yarnpkg.com/configuration/manifest/)
```
Comment (brunoapimentel) [inline README.md:371]: We do parse the `package.json` file, but only to identify the yarn version. I think the text in the gomod block is more accurate: https://github.com/eskultety/cachi2/blob/yarn-docs/README.md?plain=1#L313-L315 

I don't really like that it says "we parse the file" and in the next sentence it says "we don't realy parse it", but I think it's best to acknowledge that we use the `yarn` command to get the list of dependencies.

File: docs/yarn.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    ├── agent-base-npm-6.0.2-428f325a93-8.zip
+    ├── agentkeepalive-npm-4.3.0-ac3d8e6807-8.zip
+    ├── aggregate-error-npm-3.1.0-415a406f4e-8.zip
+    ├── ansi-regex-npm-3.0.1-01f44078a3-8.zip
+...
+```
+
+### Known pitfalls
+If your repository isn't in a pristine state, i.e. you tried to run ``yarn install`` previously on
+your own without Cachi2 what may happen is that Cachi2 will assume the repository makes use of
```
Comment (brunoapimentel) [inline docs/yarn.md:146]:
```suggestion
If your repository isn't in a pristine state (i.e. you tried to run ``yarn install`` previously on
your own without Cachi2), what may happen is that Cachi2 will assume the repository makes use of
```

File: docs/yarn.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+proceeds with the actual dependency fetching, it verifies whether all [configuration
+settings](https://v3.yarnpkg.com/configuration/yarnrc) that set a path to a resource don't point
+outside of the source repository, so in order to avoid any issues reported by Cachi2 in this regard
+make sure all your project resource references are bound by the repository. Part of the analysis of
+the repository's ``.yarnrc.yml`` file is detection of plugin usage which is further explained in
+[Dealing with plugins](#dealing-with-plugins).
+
+### Dealing with Yarn Zero-Installs
+
+Yarn's PnP Zero-Installs <https://v3.yarnpkg.com/features/zero-installs/> are unsupported due to
```
Comment (brunoapimentel) [inline docs/yarn.md:50]:
```suggestion
Yarn's [PnP Zero-Installs](https://v3.yarnpkg.com/features/zero-installs/) are unsupported due to
```

File: docs/yarn.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  "path": ".",
+}
+```
+
+or more simply by just invoking:
+``cachi2 fetch-deps yarn``
+
+For complete example of how to pre-fetch dependencies, see [Pre-fetch dependencies][usage-prefetch].
+
+### Controlling Yarn's behaviour
```
Comment (brunoapimentel) [inline docs/yarn.md:108]: :gb:

**Comment by brunoapimentel** [comment]
> > I think we also want to have a complete example, following what has been done on other package managers: https://github.com/containerbuildsystem/cachi2/blob/main/docs/usage.md#usage-examples.
> 
> Added the change, however, @brunoapimentel I'd like to ask you to move your sample-nodejs-app Yarn modifications to the main cachito nodejs repo which my docs changes refer to. Also, I think you'll need to do a few cosmetic changes there so that the docs matches is, e.g. the Dockerfile.

:+1:  https://github.com/containerbuildsystem/cachi2/pull/425#issuecomment-1857400758

File: README.md (hunk 3 above)
Comment (eskultety) [reply README.md]: Reworded. How about now?

File: docs/yarn.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+settings](https://v3.yarnpkg.com/configuration/yarnrc) that set a path to a resource don't point
+outside of the source repository, so in order to avoid any issues reported by Cachi2 in this regard
+make sure all your project resource references are bound by the repository. Part of the analysis of
+the repository's ``.yarnrc.yml`` file is detection of plugin usage which is further explained in
+[Dealing with plugins](#dealing-with-plugins).
+
+### Dealing with Yarn Zero-Installs
+
+Yarn's [PnP Zero-Installs](https://v3.yarnpkg.com/features/zero-installs/) are unsupported due to
+the potentially unplugged dependencies checked into the repository which simply make it
```
Comment (brunoapimentel) [inline docs/yarn.md:51]: Wdyt about adding a link to clarify what [unplugged dependencies](https://v3.yarnpkg.com/advanced/lexicon#unplugged-package) are?

File: docs/yarn.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    ├── aggregate-error-npm-3.1.0-415a406f4e-8.zip
+    ├── ansi-regex-npm-3.0.1-01f44078a3-8.zip
+...
+```
+
+### Known pitfalls
+If your repository isn't in a pristine state (i.e. you tried to run ``yarn install`` previously on
+your own without Cachi2) what may happen is that Cachi2 will assume the repository makes use of
+[Zero-Installs](#dealing-with-yarn-zero-installs). The workaround here is simple, just run ``yarn
+cache clean`` and cachi2 will then process your repository as normal.
```
Comment (brunoapimentel) [inline docs/yarn.md:148]: I like this explanation here, but I feel like it belongs in the previous h2 (Specifying packages to process).

File: README.md (hunk 1 above)
Comment (eskultety) [reply README.md]: > Maybe we should wrap this into a nicer error? In any case, I think to keep the current ethos of Cachi2, we shouldn't allow a repo without a lockfile to be processed.

Opened https://github.com/containerbuildsystem/cachi2/pull/440 to take care of that.

**Comment by eskultety** [comment]
Depends on https://github.com/containerbuildsystem/cachi2/pull/440

**Review by brunoapimentel — APPROVED** [review]
LGTM :rocket:

**Review by ben-alkov — APPROVED** [review]

**Pull request merged**